private/bufpkg/bufmodule/bufmodulecache/cas_module_cacher.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufmodulecache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule/bufmoduleref"
+	"github.com/bufbuild/buf/private/pkg/manifest"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+const (
+	blobsDir     = "blobs"
+	commitsDir   = "commits"
+	manifestsDir = "manifests"
+)
+
+type casModuleCacher struct {
+	logger  *zap.Logger
+	baseDir string
+}
+
+var _ moduleCache = (*casModuleCacher)(nil)
+
+func (c *casModuleCacher) GetModule(
+	ctx context.Context,
+	modulePin bufmoduleref.ModulePin,
+) (_ bufmodule.Module, retErr error) {
+	moduleBasedir := normalpath.Join(c.baseDir, modulePin.Remote(), modulePin.Owner(), modulePin.Repository())
+	var manifestHexDigest string
+	if modulePinDigestEncoded := modulePin.Digest(); modulePinDigestEncoded != "" {
+		modulePinDigest, err := manifest.NewDigestFromString(modulePinDigestEncoded)
+		if err != nil {
+			return nil, err
+		}
+		manifestHexDigest = modulePinDigest.Hex()
+	} else {
+		// Attempt to look up manifest digest from commit
+		commitPath := normalpath.Join(moduleBasedir, commitsDir, modulePin.Commit())
+		manifestDigestBytes, err := os.ReadFile(commitPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return nil, storage.NewErrNotExist(commitPath)
+			}
+			return nil, err
+		}
+		manifestHexDigest = strings.TrimSpace(string(manifestDigestBytes))
+		if manifestHexDigest == "" {
+			return nil, storage.NewErrNotExist(commitPath)
+		}
+	}
+	manifestPath := normalpath.Join(moduleBasedir, manifestsDir, manifestHexDigest[:2], manifestHexDigest[2:])
+	f, err := os.Open(manifestPath)
```

Comment:
Maybe we talked about this and I forgot, but why is manifest blobs treated special? It should work to store all manifests in `blobsDir`. If we want to be more transparent about the cache on disk, we could have the manifest dir be links or just empty.

(The order of writing needs to be changed to blobs then manifest blob, but that's simple enough.)

Reply:
There will be no overlap between manifest contents and blob contents (we won't get any space savings if we move them to the same dir), and so storing in a separate directory makes it easier to navigate the cache from disk. Otherwise, this shouldn't matter much at all to the CLI.